flocker/control/_persistence.py

```diff
@@ -162,37 +164,106 @@ def upgrade_from_v3(cls, config):
+        return result
+    elif isinstance(obj, PClass):
+        result = obj._to_dict()
+        result[_CLASS_MARKER] = obj.__class__.__name__
+        return result
+    elif isinstance(obj, PMap):
+        return {_CLASS_MARKER: u"PMap", u"values": dict(obj).items()}
+    elif isinstance(obj, (PSet, PVector, set)):
+        return list(obj)
+    elif isinstance(obj, FilePath):
+        return {_CLASS_MARKER: u"FilePath",
+                u"path": obj.path.decode("utf-8")}
+    elif isinstance(obj, UUID):
+        return {_CLASS_MARKER: u"UUID",
+                "hex": unicode(obj)}
+    elif isinstance(obj, datetime):
+        if obj.tzinfo is None:
+            raise ValueError(
+                "Datetime without a timezone: {}".format(obj))
+        return {_CLASS_MARKER: u"datetime",
+                "seconds": timegm(obj.utctimetuple())}
+    return obj
+
+
+def _is_pyrsistent(obj):
+    """
+    Boolean check if an object is a pyrsistent object.
+    """
+    return isinstance(obj, (PRecord, PClass, PMap, PSet, PVector))
+
+
+_cached_dfs_serialize_cache = WeakKeyDictionary()
+
+
+def _cached_dfs_serialize(input_object):
     """
-    JSON encoder that can encode the configuration model.
-    Base encoder for version 1 configurations.
+    This serializes an input object into something that can be serialized by
+    the python json encoder.
+
+    This caches the serialization of pyrsistent objects in a
+    ``WeakKeyDictionary``, so the cache should be automatically cleared when
+    the input object that is cached is destroyed.
+
+    :returns: An entirely serializable version of input_object.
     """
-    def default(self, obj):
-        if isinstance(obj, PRecord):
-            result = dict(obj)
-            result[_CLASS_MARKER] = obj.__class__.__name__
-            return result
-        elif isinstance(obj, PClass):
-            result = obj._to_dict()
-            result[_CLASS_MARKER] = obj.__class__.__name__
-            return result
-        elif isinstance(obj, PMap):
-            return {_CLASS_MARKER: u"PMap", u"values": dict(obj).items()}
-        elif isinstance(obj, (PSet, PVector, set)):
-            return list(obj)
-        elif isinstance(obj, FilePath):
-            return {_CLASS_MARKER: u"FilePath",
-                    u"path": obj.path.decode("utf-8")}
-        elif isinstance(obj, UUID):
-            return {_CLASS_MARKER: u"UUID",
-                    "hex": unicode(obj)}
-        elif isinstance(obj, datetime):
-            if obj.tzinfo is None:
-                raise ValueError(
-                    "Datetime without a timezone: {}".format(obj))
-            return {_CLASS_MARKER: u"datetime",
-                    "seconds": timegm(obj.utctimetuple())}
-        return JSONEncoder.default(self, obj)
+    # Ensure this is a quick function for basic types:
+    if input_object is None:
+        return None
+    t = type(input_object)
+    is_pyrsistent = False
+    if (
```

Comment:
why not:

     if isinstance(input_object, (types.NoneType, str, unicode, int, long, float, bool)):
         return input_object

Instead of lines 230-242

and

    if isinstance(input_object, (dict, list, tuple)):
        obj = input_object

instead of lines 243-246

It seems clearer. If there is a performance benefit in the less clear version, I think we should put a specific comment to that.